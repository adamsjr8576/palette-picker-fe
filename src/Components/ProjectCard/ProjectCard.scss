.project-cards-section {
  display: flex;
  flex-direction: column;
  width: 30%;
  align-items: center;
  margin: 10px 0px;
}

.palette-card-div {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-around;
  width: 100%;
  &>:hover {
    cursor: pointer;
    background-color: #eceaea;
  }
}

.project-name {
  font-size: 2em;
  text-align: left;
  margin-top: 0px;
  width: 78%;
  border-bottom: solid 1px black;
  background-color: white;
  margin-bottom: 12px;
  padding-bottom: 15px;
}

h3 {
  font-size: 1.5em;
  width: 150px;
  padding: .5em;
}