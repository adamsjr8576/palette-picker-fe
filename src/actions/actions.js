export const addProject = (project) => ({
  type: 'ADD_PROJECT',
  project
})