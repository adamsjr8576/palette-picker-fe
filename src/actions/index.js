export const addNewPalette = currentPalette => ({
  type: 'ADD_PALETTE',
  currentPalette
});

export const updatePaletteLocked = hexCode => ({
  type: 'UPDATE_PALETTE_LOCKED',
  hexCode
});

export const addInitialProjects = projects => ({
  type: 'ADD_INITIAL_PROJECTS',
  projects
});

export const addPalettes = palettes => ({
  type: 'ADD_PALETTE_TO_PROJECT',
  palettes
});

export const addProject = project => ({
  type: 'ADD_PROJECT',
  project
});

export const deletePaletteFromStore = palette => ({
  type: 'DELETE_PALETTE',
  palette
});

export const deleteProjectFromStore = project => ({
  type: 'DELETE_PROJECT',
  project
});
