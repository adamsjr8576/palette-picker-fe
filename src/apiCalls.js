export const postProject = (projectName) => {
  const options = {
    method: 'POST',
    body: JSON.stringify({
      name: projectName
    }),
    headers: {
      'Content-Type': 'application/json'
    }
  }

  return fetch(process.env.REACT_APP_BACKEND_URL + '/api/v1/projects', options)
    .then(res => res.json());
}

export const getProjectById = (id) => {
  return fetch(process.env.REACT_APP_BACKEND_URL + `/api/v1/projects/${id.id}`)
    .then(res => res.json());
}

export const getProjects = () => {
  return fetch(process.env.REACT_APP_BACKEND_URL + `/api/v1/projects`)
    .then(res => res.json());
}

export const postPalette = (body) => {
  const options = {
    method: 'POST',
    body: JSON.stringify(body),
    headers: {
      'Content-Type': 'application/json'
    }
  }

  return fetch(process.env.REACT_APP_BACKEND_URL + '/api/v1/palettes', options)
    .then(res => res.json());
}

export const getPaletteById = (id) => {
  return fetch(process.env.REACT_APP_BACKEND_URL + `/api/v1/palettes/${id.id}`)
    .then(res => res.json());
}

export const deletePalette = (id) => {
  return fetch(process.env.REACT_APP_BACKEND_URL + `/api/v1/palettes/${id}`)
    .then(res => res.json())
}