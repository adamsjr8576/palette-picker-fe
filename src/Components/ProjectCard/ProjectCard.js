import React, { useEffect } from 'react';
import './ProjectCard.scss';
import { connect } from 'react-redux';
import ProjectPalette from '../ProjectPalette/ProjectPalette';
import images from '../../images/images';
import { addPalettes, deleteProjectFromStore, deletePaletteByProjectId } from '../../actions/index';
import { deleteProject, getPaletteByProjectId  } from '../../apiCalls';

export const ProjectCard = ({ name, id, addPalettes, palettes, projects, deleteProjectFromStore, deletePaletteByProjectId }) => {

  let paletteCards

  useEffect(() => {
    getPaletteByProjectId(id)
      .then(data => {
        addPalettes(data)
      })
      .catch(err => console.log(err))
  }, []);

  const removeProject = (id) => {
    const projectToDelete = projects.find(item => item.id === id)
    deleteProject(id, projectToDelete)
    .then(res => console.log(res))

    deletePaletteByProjectId(id)
    deleteProjectFromStore(projectToDelete)
  }

  if(palettes.length) {
    const filteredPalettes = palettes.filter(palette => palette.project_id === id)
    paletteCards = filteredPalettes.map(palette => {
      return(
        <ProjectPalette
          key={palette.id}
          name={palette.name}
          id={palette.id}
          colors={palette.colors}
        />
      )
    })
  }

  return(
    <section data-testid='project-id' className='project-cards-section'>
      <div className='project-title-div'>
        <button data-testid='button-id' className='project-delete-icon-btn' onClick={ () => removeProject(id) }><img className='project-delete-icon' src={images.quit}/></button>
        <h2 data-testid='project-name'  className='project-name'>{name}:</h2>
      </div>
      <div className='palette-card-div'>
        {paletteCards}
      </div>
    </section>
  )
}

export const mapStateToProps = state => ({
  palettes: state.palettes,
  projects: state.projects
});

export const mapDispatchToProps = dispatch => ({
  addPalettes: palettes => dispatch( addPalettes(palettes) ),
  deletePaletteByProjectId: id => dispatch( deletePaletteByProjectId(id) ),
  deleteProjectFromStore: project => dispatch( deleteProjectFromStore(project) )
});

export default connect(mapStateToProps, mapDispatchToProps)(ProjectCard);
