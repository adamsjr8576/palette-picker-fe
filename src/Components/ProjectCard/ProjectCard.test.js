import React from 'react';
import { ProjectCard, mapDispatchToProps, mapStateToProps } from './ProjectCard';
import { addPalettes, deleteProjectFromStore, deletePaletteByProjectId } from '../../actions/index';
import { render, fireEvent } from '@testing-library/react';
import { deleteProject, getPaletteByProjectId  } from '../../apiCalls';

jest.mock('../../apiCalls.js');

describe('ProjectCard', () => {
  let mockProps, setup;

  describe('ProjectCard Component', () => {

    beforeEach(() => {
      getPaletteByProjectId.mockImplementation(() => {
        return Promise.resolve([{
          colors: ["#c005d4", "#57a3d0", "#88b8b1", "#7bac74", "#e6488b"],
          id: 21,
          name: "Candy",
          created_at: "2020-02-11T21:07:35.955Z",
          updated_at: "2020-02-11T21:07:35.955Z",
          project_id: 2
        }]);
      })

      deleteProject.mockImplementation(() => {
        return Promise.resolve('Successful delete!');
      })
    });


    test('should match the snapshot when palettes is an empty array', () => {
      mockProps = {
        name: 'Project 2',
        id: 5,
        addPalettes: jest.fn(),
        palettes: [],
        projects: [{name: 'Project1', id: 5}, {name: 'Project2', id: 6}],
        deleteProjectFromStore: jest.fn(),
        deletePaletteByProjectId: jest.fn()
      }

      setup = () => {
        const utils = render(<ProjectCard
          name={mockProps.name}
          id={mockProps.id}
          addPalettes={mockProps.addPalettes}
          palettes={mockProps.palettes}
          projects={mockProps.projects}
          deleteProjectFromStore={mockProps.deleteProjectFromStore}
          deletePaletteByProjectId={mockProps.deletePaletteByProjectId}
        />)
        return { utils };
      }

      const { utils } = setup();
      expect(utils).toMatchSnapshot();
    });

    test('should match the snapshot when palettes is an array of palettes', () => {
      mockProps = {
        name: 'Project 2',
        id: 5,
        addPalettes: jest.fn(),
        palettes: [{
          colors: ["#c005d4", "#57a3d0", "#88b8b1", "#7bac74", "#e6488b"],
          id: 21,
          name: "Candy",
          created_at: "2020-02-11T21:07:35.955Z",
          updated_at: "2020-02-11T21:07:35.955Z",
          project_id: 2
        },
        {
          colors: ["#c005d4", "#57a3d0", "#88b8b1", "#7bac74", "#e6488b"],
          id: 23,
          name: "Hot Sauce",
          created_at: "2020-02-11T21:07:35.955Z",
          updated_at: "2020-02-11T21:07:35.955Z",
          project_id: 2
        }],
        projects: [{name: 'Project1', id: 5}, {name: 'Project2', id: 6}],
        deleteProjectFromStore: jest.fn(),
        deletePaletteByProjectId: jest.fn()
      }

      setup = () => {
        const utils = render(<ProjectCard
          name={mockProps.name}
          id={mockProps.id}
          addPalettes={mockProps.addPalettes}
          palettes={mockProps.palettes}
          projects={mockProps.projects}
          deleteProjectFromStore={mockProps.deleteProjectFromStore}
          deletePaletteByProjectId={mockProps.deletePaletteByProjectId}
        />)
        return { utils };
      }

      const { utils } = setup();
      expect(utils).toMatchSnapshot();
    });

    test('should invoke getPaletteByProjectId on render in useEffect', () => {
      mockProps = {
        name: 'Project 2',
        id: 5,
        addPalettes: jest.fn(),
        palettes: [{
          colors: ["#c005d4", "#57a3d0", "#88b8b1", "#7bac74", "#e6488b"],
          id: 21,
          name: "Candy",
          created_at: "2020-02-11T21:07:35.955Z",
          updated_at: "2020-02-11T21:07:35.955Z",
          project_id: 2
        },
        {
          colors: ["#c005d4", "#57a3d0", "#88b8b1", "#7bac74", "#e6488b"],
          id: 23,
          name: "Hot Sauce",
          created_at: "2020-02-11T21:07:35.955Z",
          updated_at: "2020-02-11T21:07:35.955Z",
          project_id: 2
        }],
        projects: [{name: 'Project1', id: 5}, {name: 'Project2', id: 6}],
        deleteProjectFromStore: jest.fn(),
        deletePaletteByProjectId: jest.fn()
      }

      setup = () => {
        const utils = render(<ProjectCard
          name={mockProps.name}
          id={mockProps.id}
          addPalettes={mockProps.addPalettes}
          palettes={mockProps.palettes}
          projects={mockProps.projects}
          deleteProjectFromStore={mockProps.deleteProjectFromStore}
          deletePaletteByProjectId={mockProps.deletePaletteByProjectId}
        />)
        return { utils };
      }

      const { utils } = setup();
      expect(getPaletteByProjectId).toHaveBeenCalledWith(5);
    });

    test('Should delete a project when button is clicked', () => {
      mockProps = {
        name: 'Project 2',
        id: 5,
        addPalettes: jest.fn(),
        palettes: [{
          colors: ["#c005d4", "#57a3d0", "#88b8b1", "#7bac74", "#e6488b"],
          id: 21,
          name: "Candy",
          created_at: "2020-02-11T21:07:35.955Z",
          updated_at: "2020-02-11T21:07:35.955Z",
          project_id: 2
        },
        {
          colors: ["#c005d4", "#57a3d0", "#88b8b1", "#7bac74", "#e6488b"],
          id: 23,
          name: "Hot Sauce",
          created_at: "2020-02-11T21:07:35.955Z",
          updated_at: "2020-02-11T21:07:35.955Z",
          project_id: 2
        }],
        projects: [{name: 'Project1', id: 5}, {name: 'Project2', id: 6}],
        deleteProjectFromStore: jest.fn(),
        deletePaletteByProjectId: jest.fn()
      }

      setup = () => {
        const utils = render(<ProjectCard
          name={mockProps.name}
          id={mockProps.id}
          addPalettes={mockProps.addPalettes}
          palettes={mockProps.palettes}
          projects={mockProps.projects}
          deleteProjectFromStore={mockProps.deleteProjectFromStore}
          deletePaletteByProjectId={mockProps.deletePaletteByProjectId}
        />)
        const header = utils.getByTestId('project-name')
        const btn = utils.getByTestId('button-id')
        return {
          header,
          btn,
          ...utils
        };
      }

      const { header, btn, utils } = setup();
      fireEvent.click(btn);
      expect(deleteProject).toHaveBeenCalledWith(5, {name: 'Project1', id: 5});
      expect(mockProps.deletePaletteByProjectId).toHaveBeenCalledWith(5);
      expect(mockProps.deleteProjectFromStore).toHaveBeenCalledWith({name: 'Project1', id: 5});
    });

    describe('mapStateToProps', () => {
      it('Should return an object with the palettes and projects', () => {
        const mockState = {
          palettes: [{
            colors: ["#c005d4", "#57a3d0", "#88b8b1", "#7bac74", "#e6488b"],
            id: 21,
            name: "Candy",
            created_at: "2020-02-11T21:07:35.955Z",
            updated_at: "2020-02-11T21:07:35.955Z",
            project_id: 2
          }],
          projects: [{name: 'Project1', id: 5}, {name: 'Project2', id: 6}],
          currentPalette: [
            {locked: false, color: '#444343'},
            {locked: false, color: '#6D6D6D'},
            {locked: false, color: '#9B9B9B'},
            {locked: false, color: '#C2C2C2'},
            {locked: false, color: '#DCDCDC'}]
        }
        const expected = {
          palettes: [{
            colors: ["#c005d4", "#57a3d0", "#88b8b1", "#7bac74", "#e6488b"],
            id: 21,
            name: "Candy",
            created_at: "2020-02-11T21:07:35.955Z",
            updated_at: "2020-02-11T21:07:35.955Z",
            project_id: 2
          }],
          projects: [{name: 'Project1', id: 5}, {name: 'Project2', id: 6}]
        }

        const mappedProps = mapStateToProps(mockState);
        expect(mappedProps).toEqual(expected);
      });
    });

    describe('mapDispatchToProps', () => {
      it('Should call dispatch with addPalettes action', () => {
        const palettes = [{
          colors: ["#c005d4", "#57a3d0", "#88b8b1", "#7bac74", "#e6488b"],
          id: 21,
          name: "Candy",
          created_at: "2020-02-11T21:07:35.955Z",
          updated_at: "2020-02-11T21:07:35.955Z",
          project_id: 2
        }];
        const mockDispatch = jest.fn();
        const actionToDispatch = addPalettes(palettes);
        const mappedProps = mapDispatchToProps(mockDispatch);

        mappedProps.addPalettes(palettes);
        expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
      });

      it('Should call dispatch with deletePaletteByProjectId action', () => {
        const id = 5;
        const mockDispatch = jest.fn();
        const actionToDispatch = deletePaletteByProjectId(id);
        const mappedProps = mapDispatchToProps(mockDispatch);

        mappedProps.deletePaletteByProjectId(id);
        expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
      });

      it('Should call dispatch with deleteProjectFromStore action', () => {
        const project = {name: 'Project1', id: 5};
        const mockDispatch = jest.fn();
        const actionToDispatch = deleteProjectFromStore(project);
        const mappedProps = mapDispatchToProps(mockDispatch);

        mappedProps.deleteProjectFromStore(project);
        expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
      });
    });
  });
});
