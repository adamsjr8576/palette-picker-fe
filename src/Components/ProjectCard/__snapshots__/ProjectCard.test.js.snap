// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectCard ProjectCard Component should match the snapshot when palettes is an array of palettes 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="project-cards-section"
        data-testid="project-id"
      >
        <div
          class="project-title-div"
        >
          <button
            class="project-delete-icon-btn"
            data-testid="button-id"
          >
            <img
              class="project-delete-icon"
              src="quit.png"
            />
          </button>
          <h2
            class="project-name"
            data-testid="project-name"
          >
            Project 2
            :
          </h2>
        </div>
        <div
          class="palette-card-div"
        />
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="project-cards-section"
      data-testid="project-id"
    >
      <div
        class="project-title-div"
      >
        <button
          class="project-delete-icon-btn"
          data-testid="button-id"
        >
          <img
            class="project-delete-icon"
            src="quit.png"
          />
        </button>
        <h2
          class="project-name"
          data-testid="project-name"
        >
          Project 2
          :
        </h2>
      </div>
      <div
        class="palette-card-div"
      />
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ProjectCard ProjectCard Component should match the snapshot when palettes is an empty array 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="project-cards-section"
        data-testid="project-id"
      >
        <div
          class="project-title-div"
        >
          <button
            class="project-delete-icon-btn"
            data-testid="button-id"
          >
            <img
              class="project-delete-icon"
              src="quit.png"
            />
          </button>
          <h2
            class="project-name"
            data-testid="project-name"
          >
            Project 2
            :
          </h2>
        </div>
        <div
          class="palette-card-div"
        />
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="project-cards-section"
      data-testid="project-id"
    >
      <div
        class="project-title-div"
      >
        <button
          class="project-delete-icon-btn"
          data-testid="button-id"
        >
          <img
            class="project-delete-icon"
            src="quit.png"
          />
        </button>
        <h2
          class="project-name"
          data-testid="project-name"
        >
          Project 2
          :
        </h2>
      </div>
      <div
        class="palette-card-div"
      />
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
